nixos/modules/services/misc/ringboard.nix

```diff
@@ -0,0 +1,79 @@
+{
+  config,
+  lib,
+  pkgs,
+  ...
+}:
+let
+  cfg = config.services.ringboard;
+in
+{
+  options.services.ringboard = {
+    x11.enable = lib.mkEnableOption "X11 support for Ringboard";
+    wayland.enable = lib.mkEnableOption "Wayland support for Ringboard";
+    x11Package = lib.mkPackageOption pkgs "ringboard" { };
+    waylandPackage = lib.mkPackageOption pkgs "ringboard-wayland" { };
+  };
+
+  config = lib.mkIf (cfg.x11.enable || cfg.wayland.enable) {
+    systemd.user.services.ringboard-server = {
+      description = "Ringboard server";
+      documentation = [ "https://github.com/SUPERCILEX/clipboard-history" ];
+      after = [ "multi-user.target" ];
+      wantedBy = [ "multi-user.target" ];
+      serviceConfig = {
+        Type = "notify";
+        Environment = "RUST_LOG=trace";
+        ExecStart = "${if cfg.x11.enable then cfg.x11Package else cfg.waylandPackage}/bin/ringboard-server";
+        Restart = "on-failure";
+        Slice = "session-ringboard.slice";
+      };
+    };
+
+    systemd.user.services.ringboard-x11 = lib.mkIf cfg.x11.enable {
+      description = "Ringboard clipboard listener (X11)";
+      documentation = [ "https://github.com/SUPERCILEX/clipboard-history" ];
+      requires = [ "ringboard-server.service" ];
+      after = [
+        "ringboard-server.service"
+        "graphical-session.target"
+      ];
+      bindsTo = [ "graphical-session.target" ];
+      wantedBy = [ "graphical-session.target" ];
+      serviceConfig = {
+        Type = "exec";
+        ExecStart = "${cfg.x11Package}/bin/ringboard-x11";
+        Restart = "on-failure";
+        Slice = "session-ringboard.slice";
+      };
+      environment.RUST_LOG = "trace";
+    };
+
+    systemd.user.services.ringboard-wayland = lib.mkIf cfg.wayland.enable {
+      description = "Ringboard clipboard listener (Wayland)";
+      documentation = [ "https://github.com/SUPERCILEX/clipboard-history" ];
+      requires = [ "ringboard-server.service" ];
+      after = [
+        "ringboard-server.service"
+        "graphical-session.target"
+      ];
+      bindsTo = [ "graphical-session.target" ];
+      wantedBy = [ "graphical-session.target" ];
+      serviceConfig = {
+        Type = "exec";
+        ExecStart = "${cfg.waylandPackage}/bin/ringboard-wayland";
+        Restart = "on-failure";
+        Slice = "session-ringboard.slice";
+      };
+      environment.RUST_LOG = "trace";
+    };
```

Comment:
I'm wondering a bit about this solution. If you were to enable both, what happens when you start either the X11 session or the wayland session? Does both services start, and then one of them fails? If so, maybe we could set `ConditionEnvironment = "XDG_SESSION_TYPE=wayland"` and `ConditionEnvironment = "XDG_SESSION_TYPE=x11"`?

Alternatively, we could go back to just using a script to choose between them at runtime, which makes the entire thing a bit less complex. It can be optimized a bit by using `exec` to replace the parent bash process and using `lib.optionalString` (or equivalent) to not include the other package in the closure when not enabled. Is there a reason you want to avoid this @SUPERCILEX?

Reply:
I think you would end up with duplicate entries. I don't have a strong opinion here, but as @magnetophon pointed out it's possible for the same user have both an X11 login and a wayland login. So maybe it's nice to be able to have both services around? Though it would be sad if users can easily shoot themselves in the foot.

Reply:
I think it's a bit weird to have the other service fail as the expected behaviour, especially with restart on failure enabled.

Setting the options i mentioned above in `serviceConfig` will tell systemd not to run either service when `XDG_SESSION_TYPE` is not set to the expected value.

However, I think it also would be an okay solution to just revert to having a single `ringboard.service` with `script` set to something.like

```nix
''
  if [ "$XDG_SESSION_TYPE" = "wayland" ]; then
    ${lib.optionalString cfg.wayland.enable "exec '${cfg.package}'/bin/ringboard-wayland"}
  else
    ${lib.optionalString cfg.x11.enable "exec '${cfg.package}'/bin/ringboard-x11"}
  fi
''
```

That should make it so that the wayland/x11 package doesn't end up in the nix store unless enabled.